core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -642,24 +642,31 @@ public boolean rename(Path src, Path dst) throws IOException {
 
     AlluxioURI srcPath = new AlluxioURI(HadoopUtils.getPathWithoutScheme(src));
     AlluxioURI dstPath = new AlluxioURI(HadoopUtils.getPathWithoutScheme(dst));
-    ensureExists(srcPath);
-    URIStatus dstStatus;
-    try {
-      dstStatus = mFileSystem.getStatus(dstPath);
-    } catch (IOException | AlluxioException e) {
-      dstStatus = null;
-    }
-    // If the destination is an existing folder, try to move the src into the folder
-    if (dstStatus != null && dstStatus.isFolder()) {
-      dstPath = dstPath.join(srcPath.getName());
-    }
     try {
       mFileSystem.rename(srcPath, dstPath);
-      return true;
-    } catch (IOException | AlluxioException e) {
+    } catch (AlluxioException e) {
+      ensureExists(srcPath);
+      URIStatus dstStatus;
+      try {
+        dstStatus = mFileSystem.getStatus(dstPath);
+      } catch (IOException | AlluxioException e2) {
+        dstStatus = null;
+      }
+      // If the destination is an existing folder, try to move the src into the folder
+      if (dstStatus != null && dstStatus.isFolder()) {
+        dstPath = dstPath.join(srcPath.getName());
+      }
+      try {
+        mFileSystem.rename(srcPath, dstPath);
```

Comment:
In the case where `dstStatus == null` or `!dstStatus.isFolder` we should not call `rename` a second time.